### PR TITLE
PartnerOfferings UI

### PR DESCRIPTION
**Pull request by OmarPadierna**
**Description**

**Issues that should be CLOSED by merge of this PR:**
* Fixes #

**Related issues that should be LINKED to this PR:**
* Connects #

### CODE REVIEW

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    @IBOutlet weak var descriptionLabel: UILabel!
+    @IBOutlet weak var productImage: UIImageView!
+    @IBOutlet weak var expandLabel: UILabel!
+
+    public weak var delegate: ExpandingCellDelegate?
+
+    public var isExpanded: Bool? {
+        didSet {
+            descriptionLabel.numberOfLines = self.isExpanded == true ? 0:3
+            expandLabel.text = self.isExpanded == true ? "Read less":"Read more"
```
Comment (benRmc) [inline traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: Space before and after the colon.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return offeringsGroup?[section].offerings.count ?? 0
+    }
+
+    public func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        return offeringsGroup?[section].title
+    }
+
+    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "optionCell", for: indexPath) as! InfoCell
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift]: Same here

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+
+        titleLabel.text = details?.title
+        descriptionLabel.text = details?.description
+        priceLabel.text = product?.price.localizedDescriptionInBaseCurrency
+
+        // Preload images
+        details?.imageUrls.forEach({ (imageURL) in
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift]: Is this necessary? If you're asyc loading the images as cells render, each invocation should be done at about the same time.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+extension PartnerOfferingDetailsViewController: OfferingsDisplayViewControllerDelegate {
+    public func offeringsDisplayViewControllerDidChangePreferredContentSize(_ controller: OfferingsDisplayViewController) {
+        optionsView.isHidden = controller.preferredContentSize.height == 0
+        optionsHeightConstraint.constant = controller.preferredContentSize.height
+        view.layoutIfNeeded()
+    }
+
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift]: remove

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingsFetchViewController.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    override open func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        switch (segue.identifier, segue.destination) {
+        case (_, let vc as OfferingsSelectViewController):
+            vc.offeringsGroup = offerings
+            vc.product = product
+            vc.delegate = self
+        case (_, let vc as ErrorViewController):
+            vc.delegate = self
+            vc.retryButtonString = "Dismiss"
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingsFetchViewController.swift]: Add a TODO in the ErrorViewController to rename retryButton and retryButtonString to actionButton and actionButtonString to keep it more generic.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsDisplayViewController.swift [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    private var isExpandedCellInGroup: [[Bool]]?
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+
+        updatePreferredContentSize()
+
+        // Initialize array of groups to keep track of expanded cells.
+        isExpandedCellInGroup = offeringsGroup.map{$0.map { (group) -> [Bool] in
+            return [Bool](repeating: false, count: group.offerings.count)
+            }}
```
[the remark this answers is not in the document]
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsDisplayViewController.swift]: Nope, that's the right indentation. Tried it with ctrl+I and that's where it is indenting

File: traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    }
+
+    private var totalPrice: Price?
+
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+
+        tableView.delegate = self
+        tableView.dataSource = self
```
[the remark this answers is not in the document]
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift]: +1

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingsFetchViewController.swift (hunk 5 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingsFetchViewController.swift]: Renamed them instead.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift (hunk 3 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift]: ...I'm not sure what you mean by "each invocation should be done at the same time". By invoking the asset manager in view did load, you're preemptively making the network request of each image and storing it in a cache. When the asset manager is called in every row, it is now loading images from the cache not necessarily from the network.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsDisplayViewController.swift [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    override open func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return offeringsGroup?[section].offerings.count ?? 0
+    }
+
+    override open func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        return offeringsGroup?[section].title
+    }
+
+    override open func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "optionCell", for: indexPath) as! ExpandingCell
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsDisplayViewController.swift]: use the class definition for the identifier

File: traveler-ios-ui/TravelerKitUI/Storyboards/Parking.storyboard [hunk 9]
```diff
@@ -387,11 +387,11 @@
             <objects>
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Storyboards/Parking.storyboard]: Revert the changes to this file, no actual changes here

File: traveler-ios-ui/TravelerKitUI.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -280,6 +286,8 @@
 		A996650E23442846007EBF14 /* Authenticator.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = Authenticator.swift; sourceTree = "<group>"; };
 		A9D1BCFD237C5F390053C62C /* OrderPaymentViewController.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = OrderPaymentViewController.swift; sourceTree = "<group>"; };
 		A9ECCEEA225658EA00CB5E59 /* Pass+TravelerUI.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Pass+TravelerUI.swift"; sourceTree = "<group>"; };
+		CA0F66902394275300A104BE /* PartnerOfferingQuantityViewController.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = PartnerOfferingQuantityViewController.swift; sourceTree = "<group>"; };
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI.xcodeproj/project.pbxproj]: Make a group for all the PartnerOffering related controllers

File: traveler-ios-ui/TravelerKitUI/ModelExtensions/PartnerOffering+TravelerUI.swift [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+//  TravelerKitUI
+//
+//  Created by Omar Padierna on 2019-12-04.
+//  Copyright © 2019 GuestLogix. All rights reserved.
+//
+
+import Foundation
+import TravelerKit
+
+extension PartnerOffering {
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ModelExtensions/PartnerOffering+TravelerUI.swift]: Suggestion: If we don't want to extend NSMutableAttributedString, we should have a common util with static functions that we can call to do this instead.  i.e. GLXStringUtil.htmlAttributedString(...).  We keep rewriting the same thing over and over again.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        if let currentItem = offeringsGroup?[indexPath.section].offerings[indexPath.row], let group = offeringsGroup?[indexPath.section] {
+            let priceOffset = currentItem.price.valueInBaseCurrency - group.startingPrice.valueInBaseCurrency
+            cell.valueLabel.text = priceOffset > 0 ? "+\(Price(floatLiteral: priceOffset).localizedDescriptionInBaseCurrency!)" : "Included"
+        }
+
+        cell.titleLabel.text = offeringsGroup?[indexPath.section].offerings[indexPath.row].name
+        cell.accessoryType = indexPath.row == selectedItemIndexes[indexPath.section] ? .checkmark:.none
+        return cell
+    }
+
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/OfferingsSelectViewController.swift]: remove extra newline

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            vc.product = product
+            vc.delegate = self
+        default:
+            Log("Unknown segue", data: nil, level: .warning)
+        }
+    }
+}
+
+extension PartnerOfferingDetailsViewController: UICollectionViewDataSource {
+
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingDetailsViewController.swift]: remove extra newline

File: traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingQuantityViewController.swift [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    @IBOutlet weak var priceLabel: UILabel!
+    
+    var product: PartnerOfferingItem?
+    var options: [PartnerOffering]?
+    var totalPrice: Price?
+    
+    weak var delegate: PartnerOfferingQuantityViewControllerDelegate?
+
+    private var purchaseForm: PurchaseForm?
+
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/PartnerOfferingQuantityViewController.swift]: remove extra newline

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+//
+
+import UIKit
+
+public protocol ExpandingCellDelegate: class {
+    func expandingCellDidChangeContentSize(_ controller: ExpandingCell)
+}
+
+open class ExpandingCell: UITableViewCell {
+
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: remove extra newline

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    @IBOutlet weak var descriptionLabel: UILabel!
+    @IBOutlet weak var productImage: UIImageView!
+    @IBOutlet weak var expandLabel: UILabel!
+
+    public weak var delegate: ExpandingCellDelegate?
+
+    public var isExpanded: Bool? {
+        didSet {
+            descriptionLabel.numberOfLines = self.isExpanded == true ? 0 : 3
+            expandLabel.text = self.isExpanded == true ? "Read less" : "Read more"
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: self.isExpanded ? ... , No need for the == true

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var descriptionLabel: UILabel!
+    @IBOutlet weak var productImage: UIImageView!
+    @IBOutlet weak var expandLabel: UILabel!
+
+    public weak var delegate: ExpandingCellDelegate?
+
+    public var isExpanded: Bool? {
+        didSet {
+            descriptionLabel.numberOfLines = self.isExpanded == true ? 0 : 3
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: same

File: traveler-ios-ui/TravelerKitUI/ModelExtensions/PartnerOffering+TravelerUI.swift (hunk 11 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/ModelExtensions/PartnerOffering+TravelerUI.swift]: Added todo comment, I'm having @benRmc document all our TODOs into a confluence page so that we can revisit them after xmas.

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift (hunk 16 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: It is needed because isExpanded is optional (possible results: nil, true, false). Making isExpanded non-optional would require the cell to have initializers which can't be used with dequeueReusableCell.

File: traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift (hunk 17 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI/Views/ExpandingCell.swift]: see above

File: traveler-ios-ui/TravelerKitUI.xcodeproj/project.pbxproj (hunk 10 above)
Comment (OmarPadierna) [reply traveler-ios-ui/TravelerKitUI.xcodeproj/project.pbxproj]: Let's do that later, I don't wanna cause merge problems with the PR that did the grouping.
